symbolic-il2cpp/src/usym.rs

```diff
@@ -144,29 +144,46 @@ mod raw {
     pub(super) struct SourceRecord {
         /// Instruction pointer address, relative to base address of assembly.
         pub(super) address: u64,
-        /// Managed symbol name, as an offset into the strings section.
-        pub(super) symbol: u32,
-        /// Managed source file, as an offset into the strings section.
-        pub(super) file: u32,
-        /// Managed line number.
-        pub(super) line: u32,
-        // These might not even be u64, it's just 128 bits we don't know.
-        _unknown0: u64,
-        _unknown1: u64,
+        /// Native symbol name, as an offset into the strings section.
+        pub(super) native_symbol: u32,
+        /// Native source file, as an offset into the strings section.
+        pub(super) native_file: u32,
+        /// Native line number.
+        pub(super) native_line: u32,
+        /// Managed code symbol name, an offset into the strings section.
+        ///
+        /// 0 if code does not map to managed code.
+        pub(super) managed_symbol: u32,
+        /// Managed code file name, an offset into the strings section.
+        ///
+        /// 0 if code does not map to managed code.
+        pub(super) managed_file: u32,
+        /// Managed code line number, 0 if code does not map to managed code.
+        pub(super) managed_line: u32,
+        pub(super) _unknown: u32,
     }
 }
 
 /// A record mapping an IL2CPP instruction address to managed code location.
+///
+/// Not all native code maps back to managed code, for those records the managed info will
+/// be `None`.
 #[derive(Debug, Clone)]
 pub struct UsymSourceRecord<'a> {
     /// Instruction pointer address, relative to the base of the assembly.
     pub address: u64,
+    /// Symbol name of the native code.
+    pub native_symbol: Cow<'a, str>,
+    /// File name of the native code.
+    pub native_file: Cow<'a, str>,
+    /// Line number of the native code.
+    pub native_line: u32,
     /// Symbol name of the managed code.
-    pub symbol: Cow<'a, str>,
+    pub managed_symbol: Option<Cow<'a, str>>,
     /// File name of the managed code.
-    pub file: Cow<'a, str>,
+    pub managed_file: Option<Cow<'a, str>>,
     /// Line number of the managed code.
-    pub line: u32,
+    pub managed_line: Option<u32>,
```

Comment:
FWIW part of me is irked by three fields which are all `Some` or `None` together.  The over-engineering part of me wants to create something like:
```
pub struct UsymLineRecord<'a> {
    pub symbol: Cow<'a, str>,
    pub file: Cow<'a, str>,
    pub line: u32,
}
pub struct UsymSourceRecord<'a> {
    pub address: u64,
    pub native: UsymLineRecord<'a>,
    pub managed: Option<UsymLineRecord<'a>>,
}
```
And I really can't make up my mind if that's sane or not.

Reply:
I agree that it would be nice, but we should make sure that the fields really are all present or all absent.

Reply:
i'd be fine with adding something that makes this distinction obvious in a follow-up PR; it would help with the symcache conversion greatly as well.